crates/sdk/src/masp.rs

```diff
@@ -1640,235 +1628,305 @@ impl<U: ShieldedUtils + MaybeSend + MaybeSync> ShieldedContext<U> {
-                    epoch,
-                )
-                .await?;
-            // Commit the notes found to our transaction
-            for (diversifier, note, merkle_path) in unspent_notes {
-                builder
-                    .add_sapling_spend(sk, diversifier, note, merkle_path)
-                    .map_err(builder::Error::SaplingBuild)?;
-            }
-            // Commit the conversion notes used during summation
-            for (conv, wit, value) in used_convs.values() {
-                if value.is_positive() {
-                    builder
-                        .add_sapling_convert(
-                            conv.clone(),
-                            *value as u64,
-                            wit.clone(),
-                        )
-                        .map_err(builder::Error::SaplingBuild)?;
-                }
+        for MaspTransferData {
+            source,
+            target,
+            token,
+            amount,
+        } in data
+        {
+            let spending_key = source.spending_key();
+            let payment_address = target.payment_address();
+            // No shielded components are needed when neither source nor
+            // destination are shielded
+            if spending_key.is_none() && payment_address.is_none() {
+                return Ok(None);
             }
-        } else {
-            // We add a dummy UTXO to our transaction, but only the source of
-            // the parent Transfer object is used to validate fund
-            // availability
-            let source_enc = source
-                .address()
-                .ok_or_else(|| {
-                    Error::Other(
-                        "source address should be transparent".to_string(),
-                    )
-                })?
-                .serialize_to_vec();
-
-            let hash = ripemd::Ripemd160::digest(sha2::Sha256::digest(
-                source_enc.as_ref(),
-            ));
-            let script = TransparentAddress(hash.into());
-            for (digit, asset_type) in
-                MaspDigitPos::iter().zip(asset_types.iter())
+            // We want to fund our transaction solely from supplied spending key
+            let spending_key = spending_key.map(|x| x.into());
             {
-                let amount_part = digit.denominate(&amount.amount());
-                // Skip adding an input if its value is 0
-                if amount_part != 0 {
-                    builder
-                        .add_transparent_input(TxOut {
-                            asset_type: *asset_type,
-                            value: amount_part,
-                            address: script,
-                        })
-                        .map_err(builder::Error::TransparentBuild)?;
-                }
+                // Load the current shielded context given the spending key we
+                // possess
+                let mut shielded = context.shielded_mut().await;
```

Comment:
We actually want to load the context only if the source is a spending key, which we can know only from within the loop. But I've added a boolean in 937a972e40b976b1f771daa5097bef7a1c2f75f2 to avoid loading the context more than once